src/Microsoft.OData.Edm/Schema/EdmModelBase.cs

```diff
@@ -154,13 +160,29 @@ public IEnumerable<IEdmOperation> FindDeclaredOperations(string qualifiedName)
         /// <returns> A set of operations that share the binding type or empty enumerable if no such operation exists. </returns>
         public virtual IEnumerable<IEdmOperation> FindDeclaredBoundOperations(IEdmType bindingType)
         {
-            foreach (IEnumerable<IEdmOperation> operations in this.functionDictionary.Values.Distinct())
+            IList<IEdmOperation> bindableOperations;
+
+            if (!this.bindableOperationsCache.TryGetValue(bindingType.FullTypeName(), out bindableOperations))
             {
-                foreach (IEdmOperation operation in operations.Where(o => o.HasEquivalentBindingType(bindingType)))
+                HashSet<IEdmOperation> operationList = new HashSet<IEdmOperation>();
+                bindableOperations = new List<IEdmOperation>();
+
+                foreach (IList<IEdmOperation> operations in this.functionDictionary.Values)
                 {
-                    yield return operation;
+                    for (int i = 0; i < operations.Count; i++)
+                    {
+                        if (!operationList.Contains(operations[i]) && operations[i].HasEquivalentBindingType(bindingType))
+                        {
+                            bindableOperations.Add(operations[i]);
+                            operationList.Add(operations[i]);

```

Comment:
Move this check outside of loop:

    if (operationList.TryAdd(operations[i])
    {
        for (int i = 0; i < operations.Count; i++)
        {
            if (!operationList.Contains(operations[i]) && operations[i].HasEquivalentBindingType(bindingType))
            {
                bindableOperations.Add(operations[i]);
             }
    }

Reply:
Sorry didnt get that, use tryadd outside loop? operations[i] will be there only inside loop right? or am i missing something

Reply:
Sorry; typo. I meant: operationList.TryAdd(operations):

if (operationList.TryAdd(operations))
{
    for (int i = 0; i < operations.Count; i++)
    {
        if (!operationList.Contains(operations[i]) && operations[i].HasEquivalentBindingType(bindingType))
        {
            bindableOperations.Add(operations[i]);
         }
} 

I believe this has the same effect as the previous Distinct().



---
In reply to: [491743298](https://github.com/OData/odata.net/pull/1888#discussion_r491743298) [](ancestors = 491743298)